tools/pubsys/src/aws/validate_ami/mod.rs

```diff
@@ -0,0 +1,799 @@
+    info!("Parsing Infra.toml file");
+
+    // If a lock file exists, use that, otherwise use Infra.toml
+    let infra_config = InfraConfig::from_path_or_lock(&args.infra_config_path, false)
+        .context(error::ConfigSnafu)?;
+
+    let aws = infra_config.aws.clone().unwrap_or_default();
+
+    trace!("Parsed infra config: {:#?}", infra_config);
+
+    // Parse the expected ami file
+    info!("Parsing expected ami file");
+    let expected_images = parse_expected_amis(&validate_ami_args.expected_amis_path).await?;
+
+    info!("Parsed expected ami file");
+
+    // Create a HashMap of AmiClients, one for each region where validation should happen
+    let base_region = &Region::new(aws.regions[0].clone());
+    let mut ami_clients = HashMap::with_capacity(expected_images.len());
+
+    for region in expected_images.keys() {
+        let client_config = build_client_config(region, base_region, &aws).await;
+        let ami_client = AmiClient::new(&client_config);
+        ami_clients.insert(region.clone(), ami_client);
+    }
+
+    // Create a map of image_id to a bool indicating whether or not the image is public. This map
+    // is needed to determine if launch permissions should be retrieved for the image (which is the
+    // case if it is not public)
+    let expected_image_public = expected_images
+        .iter()
+        .flat_map(|(_, expected_images)| {
+            expected_images
+                .iter()
+                .map(move |image_def| (image_def.id.clone(), image_def.public))
+        })
+        .collect::<HashMap<String, bool>>();
+
+    // Retrieve the EC2 images using the AmiClients
+    info!("Retrieving EC2 images");
+    let images = describe_images(
+        &ami_clients,
+        &expected_images
+            .iter()
+            .map(|(region, images)| {
+                (
+                    region.clone(),
+                    images.iter().map(|i| i.id.clone()).collect::<Vec<String>>(),
+                )
+            })
+            .collect::<HashMap<Region, Vec<String>>>(),
+        &expected_image_public,
+    )
+    .await;
+
+    // Validate the retrieved EC2 images per region
+    info!("Validating EC2 images");
+    let results: HashMap<Region, ami::Result<HashSet<AmiValidationResult>>> = images
+        .into_iter()
+        .map(|(region, region_result)| {
+            (
+                region.clone(),
+                region_result.map(|result| {
+                    validate_images_in_region(
+                        expected_images
+                            .get(region)
+                            .map(|e| e.to_owned())
+                            .unwrap_or(vec![]),
+                        &result,
+                        region,
+                    )
+                }),
+            )
+        })
+        .collect::<HashMap<Region, ami::Result<HashSet<AmiValidationResult>>>>();
+
+    let validation_results = AmiValidationResults::new(results);
```

Comment:
nit, but usually `new` isn't appropriate when you're initializing something that's computation heavy like that. A better name would be `from_result_map` or `evaluate_results()` or something.

```suggestion
    let validation_results = AmiValidationResults::new(results);
```